### ISSUE TITLE
IOException while executing process with args

### ISSUE THREAD

**Issue by loo5mity** [github/codeql#15649]
Hello, I got following error when create database,

```
sudo arch -x86_64 codeql database create /Users/a/CodeQL/databases/micro_service_seclab --language="java" --command="mvn clean install -DskipTests --file pom.xml"  --source-root=/Users/a/CodeQL/micro_service_seclab --overwrite

Initializing database at /Users/a/CodeQL/databases/micro_service_seclab.
Running build command: [mvn, clean, install, -DskipTests, --file, pom.xml]
A fatal error occurred: IOException while executing process with args: [/Users/a/CodeQL/codeql-cli/tools/osx64/preload_tracer, mvn, clean, install, -DskipTests, --file, pom.xml]
(eventual cause: IOException "error=13, Permission denied")
```

**Comment by ginsbach** [comment]
Can you confirm that the `mvn` binary on your system is executable?

I.e. running `mvn clean install -DskipTests --file pom.xml` directly (rather than through the `codeql database create` command) does not cause an error?

**Comment by aibaars** [comment]
Coud you also check the permission on `/Users/a/CodeQL/codeql-cli/tools/osx64/preload_tracer` ?

**Comment by loo5mity** [comment]
> Could you also check the permission on `/Users/a/CodeQL/codeql-cli/tools/osx64/preload_tracer` ?

I executed the `chmod`  command
```
-rwxr-xr-x@ 1 a  staff  xxxx Jan  1  1980 preload_tracer
```
![error](https://github.com/github/codeql/assets/37500776/1fb95f16-be6b-4296-9319-d6abe0d30980)



Then the following error occurred through the codeql database create command
```
Unable to open 'preload_tracer' because the developer cannot be authenticated.
macOS cannot verify whether this app contains malware.
```

**Comment by aibaars** [comment]
I wonder why the `x` permission was missing. I also wonder why the binary has an `@` sign. Could you run `xattr -l  /Users/a/CodeQL/codeql-cli/tools/osx64/preload_tracer` to display any extended file attributes? If I recall correctly macOs puts a special attribute on things that are put in the Downloads folder to prevent running downloaded stuff by mistake. Perhaps something similar is happening here. Could you also run `codesign -dv --verbose=4` on the binary to show the signing information. 

Finally which version of macOS are you running and which version of CodeQL are you using?

@AlexDenisov any ideas what could be causing the code signing problem?

**Comment by loo5mity** [comment]
```
xattr -l  ./preload_tracer
com.apple.quarantine: 0082;65d1bd81;Keka;

codesign -dv -verbose=4 ./preload_tracer
./preload_tracer: code object is not signed at all
```

version of macOS:14.1.1
the latest version of CodeQL  :2.16.2

**Comment by aibaars** [comment]
Could you try downloading CodeQL 2.16.2 again from https://github.com/github/codeql-cli-binaries/releases ? I just downloaded a copy myself and got the following:
```
codesign -dv --verbose=4  /tmp/codeql/tools/osx64/preload_tracer 
Executable=/private/tmp/codeql/tools/osx64/preload_tracer
Identifier=libtrace_fuzz_tester-arm64.out
Format=Mach-O universal (x86_64 arm64)
CodeDirectory v=20400 size=8056 flags=0x20002(adhoc,linker-signed) hashes=248+0 location=embedded
VersionPlatform=1
VersionMin=720896
VersionSDK=918016
Hash type=sha256 size=32
CandidateCDHash sha256=2210527dedba9375610959ef745c2725e718a31b
CandidateCDHashFull sha256=2210527dedba9375610959ef745c2725e718a31b45b26315b385691780dc5486
Hash choices=sha256
CMSDigest=2210527dedba9375610959ef745c2725e718a31b45b26315b385691780dc5486
CMSDigestType=2
Executable Segment base=0
Executable Segment limit=655360
Executable Segment flags=0x1
Page size=4096
Launch Constraints:
	None
CDHash=2210527dedba9375610959ef745c2725e718a31b
Signature=adhoc
Info.plist=not bound
TeamIdentifier=not set
Sealed Resources=none
Internal requirements=none
```

**Comment by aibaars** [comment]
From the above file attribute, it looks like  the file was unarchived using https://www.keka.io/en/, and put in quarantine by the OS.

Could you try the [Extract the zip archive](https://docs.github.com/en/code-security/codeql-cli/getting-started-with-the-codeql-cli/setting-up-the-codeql-cli#2-extract-the-zip-archive) instructions? Unzipping an archive containing binaries can be a little tricky on MacOS.

**Comment by loo5mity** [comment]
Thank you, my problem has been solved.

I downloaded CodeQL 2.16.2 from https://github.com/github/codeql-cli-binaries/releases and made a zipped backup of the file.

When I ran it, I found that the file did not have executable permissions. After modifying it, it was still wrong, so I restored it from the backup file.So the file attributes are not consistent.

The download file may be incomplete due to network reasons and does not have executable permissions.After I tried downloading a few more times, the file had executable permissions.

Now I can execute `codeql database create`  command successfully.
